marklogic-data-hub/src/main/java/com/marklogic/hub/flow/impl/FlowRunnerImpl.java

```diff
@@ -222,8 +224,9 @@ public void run() {
                 }
 
                 Job stepResp = stepRunner.run();
-                stepOutputs.put(stepNum, stepResp);
                 stepRunner.awaitCompletion();
+                stepOutputs.put(stepNum, stepResp);
+
```

Comment:
https://github.com/marklogic/marklogic-data-hub/blob/423a73efb7ad178b25fd088458df1abcbf52196b/marklogic-data-hub/src/main/java/com/marklogic/hub/flow/impl/FlowRunnerImpl.java#L242-L245

I think we do not need to submit the thread again called in initializeFlow. We can do sth like:

Also I feel we do not need to peek from jobQueue. And remove it later. And you can use: jobQueue.remove() rather than jobQueue.remove(jobQueue.peek());

            while (!jobQueue.isEmpty()) {
                Queue<String> stepQueue = stepsMap.get(jobQueue.remove());
                while (!stepQueue.isEmpty()) {
                    //...
                }
            }

Reply:
1. I included jobQueue.remove() in the latest commit.
2. FlowRunnerTask is meant to run a single flow, that's why I create another Runnable and submit to the executor service in case if the queue is not empty

Reply:
We can discuss tomorrow.  My point is that we can let the available jobs in jobQueue run one by one (outer loop) in the thread until the thread exits, and we do not need to submit the runnable again in the run method.

Reply:
Also why do we need to threadPool.shutdownNow() for each flowrunner thread? we can use a shutdownHook thread to manage the pool when jvm exits.